### DISCUSSION TITLE
Computing half-light radius and ellipticity for double Sersic profile

### DISCUSSION THREAD

**Discussion by ajshajib** [AstroBridge/BDLensing#253] [General]
The code below shows an example of how to convert a parameter array from the MCMC chain into a keyword dictionary similar to `kwargs_result`.  It will also solve the linear inversion to fill in the appropriate `amp` values for the given set of non-linear parameters.

```
from lenstronomy.Sampling.parameters import Param
from lenstronomy.Util.class_creator import create_im_sim


param = Param(kwargs_model,
                      kwargs_params['lens_model'][2],
                      kwargs_params['source_model'][2],
                      kwargs_params['lens_light_model'][2],
                      kwargs_params['point_source_model'][2],
                      kwargs_lens_init=kwargs_params['lens_model'][0],
                      **kwargs_constraints
                      )

# `args` is the array of parameter set from the MCMC chain
kwargs = param.args2kwargs(args)

im_sim = create_im_sim(multi_band_list_out,
                                        'multi-linear',
                                        kwargs_model,
                                        bands_compute=None,
                                        likelihood_mask_list=likelihod_mask_list)

im_sim.image_linear_solve(**kwargs, inv_bool=True)
# after this line, `kwargs` will have the `amp` values filled in
```

For a double Sersic profile, the following code computes the half-light radius and effective ellipticity parameters:

```
from lenstronomy.Analysis.light_profile import LightProfileAnalysis
from lenstronomy.LightModel.light_model import LightModel


light_analysis = LightProfileAnalysis(LightModel(lens_light_model_list))

model_bool_list = ... 
# list of booleans with the length of `lens_light_model_list`, True for light profiles of 
# the main deflector galaxy, False for light profiles of other galaxies

r_eff = light_analysis.half_light_radius(kwargs['kwargs_lens_light'], 
                                            grid_spacing=pixel_size, 
                                            grid_num=pixel_num,
                                            model_bool_list=model_bool_list)

e1, e2 = light_analysis.ellipticity(kwargs['kwargs_lens_light'], 
                                            grid_spacing=pixel_size, 
                                            grid_num=pixel_num,
                                            model_bool_list=model_bool_list)
```

**Comment by ajshajib** [comment]
@aaimtiaz @SulymanHossain

**Reply by SulymanHossain** [reply]
@ajshajib, this code is only necessary when the ellipticity parameters between two Sersic profiles are disjointed, right?

**Reply by ajshajib** [reply]
Yes.

**Reply by aaimtiaz** [reply]
Thank you vai.

**Reply by SulymanHossain** [reply]
Just using directly this line `light_analysis = LightProfileAnalysis(lens_light_model_list)` is giving me an error:

![R_eff_error](https://github.com/ajshajib/BDLensing/assets/132705545/b2671743-1e1c-4f72-b2ae-b86dd732d0c2)
But it fixes when I provide `lens_light_model_list` through the `LightModel` instance as:
`light_model = LightModel(lens_light_model_list)`
`light_analysis = LightProfileAnalysis(light_model)`

It's correct, right?

**Reply by ajshajib** [reply]
Yes, you are right; thanks for catching this! I will fix the code above.

**Comment by aaimtiaz** [comment]
Haven't understood this line of code:
Imean, what should be the args?
```
# `args` is the array of parameter set from the MCMC chain
kwargs = param.args2kwargs(args)
```

**Reply by ajshajib** [reply]
When you don't understand what a variable value should be, looking at the docstring of the function that takes in the variable as input to find more information about it can usually be helpful. Have you tried that?

**Reply by ajshajib** [reply]
Following up on this, `args` should be the position of a walker from the MCMC chain, so it's a 1D array with length equal to the number of parameters in the MCMC chain.

**Comment by SulymanHossain** [comment]
@ajshajib bhaia, using this method and extracting data from shortened_samples_mcmc(which is after convergence), I have calculated R_eff_light sample. But this sample has almost no variation which gives zero uncertainties. Will this help in our analysis?

**Reply by ajshajib** [reply]
Almost no variation, or no variation at all? If the former, then try making the pixel size in `light_analysis.half_light_radius()` smaller. If the uncertainty is smaller than the pixel size, then you won't see it. If the latter, then have you verified that the input `kwargs` are not the same?

**Reply by SulymanHossain** [reply]
> Almost no variation, or no variation at all? If the former, then try making the pixel size in `light_analysis.half_light_radius()` smaller. If the uncertainty is smaller than the pixel size, then you won't see it. If the latter, then have you verified that the input `kwargs` are not the same?

It's the former case, and I also verified 'kwargs' changes with each iteration. Okay, now I will try with a lower pixel size.

**Comment by rafee-adnan** [comment]
We are just considering the double sersic profile for the calculation of the `R_eff_light` of the central lens galaxy for the final deliverables, the sersic profile for satellite galaxy will not be included in the calculation. Right? @ajshajib

**Reply by ajshajib** [reply]
Yes.
